lib/async_hooks.js

```diff
@@ -223,31 +223,23 @@ const storageHook = createHook({
 class AsyncLocalStorage {
   constructor() {
     this.kResourceStore = Symbol('kResourceStore');
-    this.enabled = false;
   }
 
   disable() {
-    if (this.enabled) {
-      this.enabled = false;
-      // If this.enabled, the instance must be in storageList
-      storageList.splice(storageList.indexOf(this), 1);
-      if (storageList.length === 0) {
-        storageHook.disable();
-      }
+    storageList.splice(storageList.indexOf(this), 1);
+    if (storageList.length === 0) {
+      storageHook.disable();
     }
   }
 
   // Propagate the context from a parent resource to a child one
   _propagate(resource, triggerResource) {
     const store = triggerResource[this.kResourceStore];
-    if (this.enabled) {
-      resource[this.kResourceStore] = store;
-    }
+    resource[this.kResourceStore] = store;
```

Comment:
Would it make sense to add an `if (store != null)` here to avoid propagating undefined and adding unneeded properties on resource?

Reply:
Maybe `if (store != undefined)` as the retrieval would be functionally identical? One might actually _want_ to store a `null` for some reason--could be used as a sentinel value to differentiate between a happy state, a disabled state and a middle state. :thinking:

Reply:
IMO `resource[this.kResourceStore] = store;` is necessary even in case of `undefined` or `null` values of `store`. That's because async resources may be reused and we need to ensure that the old store is properly overwritten.

Reply:
Async resources should _not_ be used. If they are, that's a bug.

Reply:
As it was discussed in #30959, [docs](https://github.com/nodejs/node/blob/master/doc/api/async_hooks.md#resource) allow userland modules to reuse async resources. So, even if core modules are not a problem, this argument is strong enough to keep this assignment as is, at least for now.

Reply:
This reuse is something we have to work on anyway. I have one PR in preparation to fix the last reuse I found in core.
But I'm quite sure that we have to work on napi - even if it is just to document that reuse is not allowed.

Reply:
That'd be a doc issue then because it's _definitely_ not _intended_ that resources should be reused. If a handle of some sort _is_ being reused, it should be creating an `AsyncResource` for each reuse and using _that_ as the `async_hooks` resource. _Any_ reuse at all of resources given to `async_hooks` is 100% a bug as it will break `executionAsyncResource`.

Reply:
Have we come to any agreement on what (if any?) the task is here?

Reply:
At some other place we discussed allowing `undefined | null` as store and the agreement was to allow it. Therefore we should propagate it also here.
Or in other words: Keep it as it is